[Select] Add generic support for value

AntoineGrandchamp opened the pull request:
It adds the opportunity to type the value of the Select component.

- [x] I have followed (at least) the [PR section of the contributing guide](https://github.com/mui-org/material-ui/blob/HEAD/CONTRIBUTING.md#sending-a-pull-request).

eps1lon commented on packages/material-ui/src/Select/SelectInput.d.ts:
```diff
@@ -1,35 +1,35 @@
 }
 
-declare const SelectInput: React.ComponentType<SelectInputProps>;
+declare const SelectInput: React.ComponentType<SelectInputProps<any>>;
```
Should use the default value.

eps1lon commented on packages/material-ui/src/Select/Select.d.ts:
```diff
@@ -142,7 +142,7 @@ export interface SelectProps
    * @returns {ReactNode}
    */
-  renderValue?: (value: SelectProps['value']) => React.ReactNode;
+  renderValue?: (value: SelectProps<T>['value']) => React.ReactNode;
```
Would this work now?
```suggestion
  renderValue?: (value: T) => React.ReactNode;
```

eps1lon reviewed (changes requested):
Thanks for the PR. Please add some tests including potential problems of this approach (e.g. how you can produce invalid typings by using a different value type in the children).

AntoineGrandchamp reviewed (commented):
It is my first time contributing to this project. Where should I add the tests you are talking about?

eps1lon commented:
> It is my first time contributing to this project. Where should I add the tests you are talking about?

`packages/material-ui/src/Select/Select.spec.tsx` would be ideal. Check out existing `*.spec.tsx?` files for examples on how to test types e.g. https://github.com/mui-org/material-ui/blob/next/packages/material-ui/src/Tabs/Tabs.spec.tsx

AntoineGrandchamp commented:
I added the generic type to `value` and `defaultValue`, that was missing. I also added some tests for `value`, `defaultValue` and `onChange`. I have not found how to enforce the type on the children.

Let me know what you think of it.

eps1lon reviewed (changes requested):
The tests should include that the generic argument is not necessarily sound e.g.

```tsx
<Select onChange={(event, value) => console.log(value} value="1">
  <MenuItem value="1" />
  // Whoops. The value in onChange won't be a string
  <MenuItem value={2} />
</Select>
```

And some inference tests e.g. `<Select value="1" onChange={(event, value) => { /* what is the type of `value`? */}}`.

AntoineGrandchamp commented:
I added a test for inference. I have no idea of how to match the `<Select>`'s value type with the `<MenuItem>`'s value type, though. The `children` of `<Select>` can be any `React.ReactNode`. I am far from being an expert of Typescript, but I want to make this work. If you have any tips, it would be appreciated.

eps1lon commented:
> I added a test for inference. I have no idea of how to match the <Select>'s value type with the <MenuItem>'s value type, though. The children of <Select> can be any React.ReactNode. I am far from being an expert of Typescript, but I want to make this work. If you have any tips, it would be appreciated.

You can't. That is the point of the tests. It's just about codifying the current limitations of TypeScript: You can make it look like it's sound with generics but it never will be.

AntoineGrandchamp commented:
> You can't. That is the point of the tests. It's just about codifying the current limitations of TypeScript: You can make it look like it's sound with generics but it never will be.

Ho, I see. I added the test you proposed.

eps1lon commented:
@AntoineGrandchamp Thanks!
